src/corelibs/U2View/src/ov_sequence/ADVGraphModel.cpp

```diff
@@ -823,188 +447,141 @@ void GSequenceGraphDrawer::showSettingsDialog() {
     }
 }
 
-bool PairVector::isEmpty() const {
-    QVector<float> emptyFp(firstPoints.size(), GSequenceGraphDrawer::UNKNOWN_VAL);
-    QVector<float> emptySp(secondPoints.size(), GSequenceGraphDrawer::UNKNOWN_VAL);
-    QVector<float> emptyCutoff(0);
-    return firstPoints == emptyFp && secondPoints == emptySp && cutoffPoints == emptyCutoff;
-}
-
-CalculatePointsTask::CalculatePointsTask(const QSharedPointer<GSequenceGraphData> &d, int numPoints, int alignedFirst, int alignedLast, bool expandMode, const GSequenceGraphWindowData &wdata, U2SequenceObject *o, const U2Region &visibleRange)
-    : BackgroundTask<PairVector>(tr("Calculate graph points"), TaskFlag_None),
-      graphUpdater(d, numPoints, alignedFirst, alignedLast, expandMode, wdata, o, visibleRange, stateInfo) {
+CalculatePointsTask::CalculatePointsTask(const QList<QSharedPointer<GSequenceGraphData>> &_graphDataList, U2SequenceObject *_sequenceObject)
+    : BackgroundTask<QList<QVector<float>>>(tr("Calculate graph points"), TaskFlag_None),
+      graphDataList(_graphDataList), sequenceObject(_sequenceObject) {
 }
 
 void CalculatePointsTask::run() {
-    graphUpdater.recalculateGraphData();
-}
-
-GraphPointsUpdater::GraphPointsUpdater(const QSharedPointer<GSequenceGraphData> &d, int numPoints, int alignedFirst, int alignedLast, bool expandMode, const GSequenceGraphWindowData &wdata, U2SequenceObject *o, const U2Region &visibleRange, U2OpStatus &os)
-    : d(d),
-      alignedFirst(alignedFirst),
-      alignedLast(alignedLast),
-      expandMode(expandMode),
-      wdata(wdata),
-      o(o),
-      visibleRange(visibleRange),
-      os(os) {
-    result.firstPoints.resize(numPoints);
-    result.firstPoints.fill(GSequenceGraphDrawer::UNKNOWN_VAL);
-    result.secondPoints.resize(numPoints);
-    result.secondPoints.fill(GSequenceGraphDrawer::UNKNOWN_VAL);
-}
-
-void GraphPointsUpdater::recalculateGraphData() {
-    CHECK(!o.isNull(), );
-
-    QVector<float> newCutoff;
-    int lastAligned = o->getSequenceLength() - o->getSequenceLength() % wdata.step;
-    U2Region r = U2Region(0, lastAligned);
-    d->ga->calculate(result.allCutoffPoints, o, r, &wdata, os);
-
-    updateGraphData();
-}
-
-void GraphPointsUpdater::updateGraphData() {
-    setChahedDataParametrs();
-
-    if (result.allCutoffPoints.isEmpty()) {
-        result.allCutoffPoints = d->cachedData.allCutoffPoints;
-    }
-    calculateCutoffPoints();
-    CHECK_OP(os, );
-    if (expandMode) {
-        calculateWithExpand();
-    } else {
-        calculateWithFit();
+    CHECK(!sequenceObject.isNull(), );
+
+    GTIMER(cvar, tvar, "GraphPointsUpdater::calculateAlgorithmPoints");
+    for (auto graphData : qAsConst(graphDataList)) {
+        try {
+            QVector<float> dataPoints;
+            graphData->algorithm->calculate(dataPoints, sequenceObject, graphData->window, graphData->step, stateInfo);
+            CHECK_OP(stateInfo, );
+            result << dataPoints;
+        } catch (const std::bad_alloc &) {
```

Comment:
`TaskScheduler` can catch this exception, not sure that it's necessary

Reply:
It won't give the same level of details in the error message and won't execute other lines below the catch block.